File: 2024/day_2.py
import utils


helper = utils.InputHelper(day=2, year=2024)
helper.fetch()


LINES: list[str] = helper.as_lines()
# LINES: list[str] = open("EXAMPLE.txt").read().strip().splitlines()


def check(nums):
    prev = None
    prev_dist = 0

    for num in nums:
        n = int(num)

        if not prev:
            prev = n
            continue

        dist = prev - n
        if abs(dist) > 3:
            return False

        if dist == 0:
            return False
        elif dist > 0 and prev_dist < 0:
            return False
        elif dist < 0 and prev_dist > 0:
            return False

        prev_dist = dist
        prev = n

    return True


P1 = 0
for line in LINES:
    nums = line.split()
    P1 += check(nums)


P2 = 0
for line in LINES:
    nums = line.split()
    P2 += check(nums) or any(check(nums[:n] + nums[n + 1 :]) for n in range(len(nums)))


print(P1)
print(P2)

File: 2024/day_2.go
package main

import (
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

// ...
func d2_check(nums []int) int {
	prev := 0
	prev_dist := 0

	for _, n := range nums {
		if prev == 0 {
			prev = n
			continue
		}

		dist := prev - n
		if absInt(dist) > 3 || dist == 0 {
			return 0
		} else if dist > 0 && prev_dist < 0 || dist < 0 && prev_dist > 0 {
			return 0
		}

		prev_dist = dist
		prev = n
	}

	return 1
}

func getInts(lines []string) [][]int {
	var inner []int
	var outter [][]int

	for _, s := range lines {
		for _, n := range strings.Split(s, " ") {
			num, _ := strconv.Atoi(n)
			inner = append(inner, num)
		}
		outter = append(outter, inner)
		inner = []int{}
	}

	return outter
}

func d2() {
	data, err := os.ReadFile("inputs/input_2.txt")
	// data, err := os.ReadFile("EXAMPLE.txt")
	if err != nil {
		panic(err)
	}

	lines := strings.Split(string(data), "\r\n")
	new := getInts(lines)

	p1 := 0
	p2 := 0

	for _, n := range new {
		p1 += d2_check(n)
	}

	for _, n := range new {
		if d2_check(n) == 1 {
			p2 += 1
			continue
		}

		for i := range len(n) {
			thonk := slices.Concat(n[0:i], n[i+1:])

			if d2_check(thonk) == 1 {
				p2 += 1
				break
			}
		}
	}

	fmt.Printf("DAY TWO TOTAL_ONE: %d\n", p1)
	fmt.Printf("DAY TWO TOTAL_TWO: %d\n", p2)
}
